examples/experiments/stlc_strategies.makam

```diff
@@ -1,67 +1,86 @@
 
-eval (app E1 E2) V <-
-  eval_strategy cbneed,
-  eval E1 (lam Body),
-  ( thunk:term ->
-    (eval thunk V2 <- memoized(eval E2 V2)) ->
-    eval (Body thunk) V ).
+thunk : term -> term -> term.
 
-tuple : list term -> term.
-eval (tuple ES) (tuple VS) <-
-  map eval ES VS.
+eval N0 (app E1 E2) V N2 when eval_strategy cbneed <-
+  eval N0 E1 (lam Body) N1,
+  (* the `V2` unification variable will be shared between all substitutions
+     of `x` within `Body`; with the evaluation rules for `thunk` that follow,
+     we can make sure that the `V2` result is re-used once it is evaluated *)
+  eval N1 (Body (thunk E2 V2)) V N2.
 
-side_effect : term.
-eval (side_effect) (tuple []) <- print "evaluation".
+eval N0 (thunk E2 V2) V2 N0 when not(refl.isunif V2).
 
-eval_cbv : term -> term -> prop.
-eval_cbnd : term -> term -> prop.
-eval_cbnm : term -> term -> prop.
-eval_cbv X Y <- (eval_strategy cbval -> eval X Y).
-eval_cbnm X Y <- (eval_strategy cbname -> eval X Y).
-eval_cbnd X Y <- (eval_strategy cbneed -> eval X Y).
+eval N0 (thunk E2 V2) V2 N1 when refl.isunif(V2) <-
+  eval N0 E2 V2 N1.
 
+tuple : list term -> term.
+eval N0 (tuple []) (tuple []) N0.
+eval N0 (tuple (X :: XS)) (tuple (Y :: YS)) N2 <-
+  eval N0 X Y N1,
+  eval N1 (tuple XS) (tuple YS) N2.
 
-(eq TERM (app (lam (fun x => tuple [x, x])) side_effect),
- print_string "cbv eval\n", eval_cbv TERM CBV_VAL, print CBV_VAL,
- print_string "cbname eval\n", eval_cbnm TERM CBNM_VAL, print CBNM_VAL,
- print_string "cbneed eval\n", eval_cbnd TERM CBND_VAL, print CBND_VAL) ?
-
+side_effect : term.
+eval N0 (side_effect) (tuple []) N1 <- plus N0 1 N1.
+
+eval_cbv : term -> term -> int -> prop.
+eval_cbnd : term -> term -> int -> prop.
+eval_cbnm : term -> term -> int -> prop.
+eval_cbv X Y N1 <- (eval_strategy cbval -> eval 0 X Y N1).
+eval_cbnm X Y N1 <- (eval_strategy cbname -> eval 0 X Y N1).
+eval_cbnd X Y N1 <- (eval_strategy cbneed -> eval 0 X Y N1).
+
+
+stlc_strategies : testsuite.
+%testsuite stlc_strategies.
+
+>> (eq _TERM (app (lam (fun x => tuple [x, x])) side_effect),
+    eval_cbv _TERM CBV_VAL CBV_COUNTER,
+    eval_cbnm _TERM CBNM_VAL CBNM_COUNTER,
+    eval_cbnd _TERM CBND_VAL CBND_COUNTER) ?
+>> Yes:
+>> CBV_VAL := tuple [ (tuple [  ]), (tuple [  ]) ],
+>> CBV_COUNTER := 1,
+>> CBNM_VAL := tuple [ (tuple [  ]), (tuple [  ]) ],
+>> CBNM_COUNTER := 2,
+>> CBND_VAL := tuple [ (tuple [  ]), (tuple [  ]) ],
+>> CBND_COUNTER := 1.
+
+
+>> (eq _TERM (app (app (lam (fun x => lam (fun y => tuple [x, x, y, y]))) side_effect) (tuple [])),
+    eval_cbv _TERM CBV_VAL CBV_COUNTER,
+    eval_cbnm _TERM CBNM_VAL CBNM_COUNTER,
+    eval_cbnd _TERM CBND_VAL CBND_COUNTER) ?
+>> Yes:
+>> CBV_VAL := tuple [ (tuple [  ]), (tuple [  ]), (tuple [  ]), (tuple [  ]) ],
+>> CBV_COUNTER := 1,
+>> CBNM_VAL := tuple [ (tuple [  ]), (tuple [  ]), (tuple [  ]), (tuple [  ]) ],
+>> CBNM_COUNTER := 2,
+>> CBND_VAL := tuple [ (tuple [  ]), (tuple [  ]), (tuple [  ]), (tuple [  ]) ],
+>> CBND_COUNTER := 1.
```

Comment:
I think it would make sense to have an example where call-by-value is _bad_

```suggestion
>> CBND_COUNTER := 1.

>> (eq _TERM (app (lam (fun x => tuple [])) side_effect),
    eval_cbv _TERM CBV_VAL CBV_COUNTER,
    eval_cbnm _TERM CBNM_VAL CBNM_COUNTER,
    eval_cbnd _TERM CBND_VAL CBND_COUNTER) ?
>> Yes:
>> CBV_VAL := tuple [  ],
>> CBV_COUNTER := 1,
>> CBNM_VAL := tuple [  ],
>> CBNM_COUNTER := 0,
>> CBND_VAL := tuple [  ],
>> CBND_COUNTER := 0.

```

_Disclaimer:_ I didn't run this

Reply:
Good idea, thanks!